nipype/algorithms/confounds.py

```diff
@@ -1201,20 +1298,56 @@ def compute_noise_components(imgseries, mask_images, num_components,
         # "The covariance matrix C = MMT was constructed and decomposed into its
         # principal components using a singular value decomposition."
         try:
-            u, _, _ = fallback_svd(M, full_matrices=False)
+            u, s, _ = fallback_svd(M, full_matrices=False)
         except np.linalg.LinAlgError:
-            if self.inputs.failure_mode == 'error':
+            if failure_mode == 'error':
                 raise
-            u = np.ones((M.shape[0], num_components), dtype=np.float32) * np.nan
+            if components_criterion >= 1:
+                u = np.full((M.shape[0], components_criterion),
+                            np.nan, dtype=np.float32)
+            else:
+                u = np.full((M.shape[0], 1), np.nan, dtype=np.float32)
+
+        variance_explained = (s ** 2) / np.sum(s ** 2)
```

Comment:
This line will still fail if `fallback_svd` fails. `s` must be defined by this point.